models/sudoswap/ethereum/sudoswap_ethereum_pools.sql

```diff
@@ -0,0 +1,275 @@
+        WHEN _poolType = 1 THEN 'nft'
+        WHEN _poolType = 2 THEN 'trade'
+      END AS pooltype,
+      _spotPrice / 1e18 AS spotprice,
+      _nft AS nftcontractaddress,
+      _initialNFTIDs AS nft_ids,
+      _fee AS initialfee,
+      output_pair AS pair_address,
+      call_block_time AS block_time,
+      tx.FROM AS creator_address,
+      date_trunc('day', now()) - date_trunc('day', call_block_time) AS days_passed,
+      cardinality(_initialNFTIDs) AS initial_nft_count,
+      tx.value / 1e18 AS initial_eth
+    FROM
+      {{ source('sudo_amm_ethereum','LSSVMPairFactory_call_createPairETH') }} cre
+      INNER JOIN {{ source('ethereum','transactions') }} tx ON tx.hash = cre.call_tx_hash
+    WHERE
+      call_success
+  ),
+  most_recent_spot_delta AS (
+    SELECT
+      COALESCE(del.contract_address, spot.contract_address) AS pair_address,
+      del.delta,
+      spot.spotprice
+    FROM
+      (
+        SELECT
+          *
+        FROM
+          (
+            SELECT
+              contract_address,
+              newDelta / 1e18 AS delta,
+              row_number() over (
+                PARTITION BY
+                  contract_address
+                ORDER BY
+                  evt_block_number DESC,
+                  evt_index DESC
+              ) as most_recent
+            FROM
+              {{ source('sudo_amm_ethereum','LSSVMPair_general_evt_DeltaUpdate') }}
+          ) a
+        WHERE
+          most_recent = 1
+      ) del
+      FULL OUTER JOIN (
+        SELECT
+          *
+        FROM
+          (
+            SELECT
+              contract_address,
+              newSpotPrice / 1e18 AS spotprice,
+              row_number() over (
+                PARTITION BY
+                  contract_address
+                ORDER BY
+                  evt_block_number DESC,
+                  evt_index DESC
+              ) AS most_recent
+            FROM
+              {{ source('sudo_amm_ethereum','LSSVMPair_general_evt_SpotPriceUpdate') }}
+          ) a
+        WHERE
+          most_recent = 1
+      ) spot ON del.contract_address = spot.contract_address
+  ),
+  erc721_balances AS (
+    SELECT
+      holder_address,
+      sum(value) AS tokens_held
+    FROM
+      (
+        SELECT
+          to AS holder_address,
+          count(evt_tx_hash) AS value
+        FROM
+          {{ source('erc721_ethereum','evt_transfer') }}
+        WHERE
+          to IN (
+            SELECT
+              pair_address
+            FROM
+              pairs_created
+          )
+        GROUP BY
+          1
+        union all
+        SELECT
+        FROM
+          AS holder_address,
+          count(evt_tx_hash) * -1 as value
+        FROM
+          {{ source('erc721_ethereum','evt_transfer') }}
+        WHERE
+        FROM
+          IN (
+            SELECT
+              pair_address
+            FROM
+              pairs_created
+          )
+        GROUP BY
+          1
+      ) a
+    GROUP BY
+      1
+  ),
```

Comment:
Should probably also filter for the correct erc721 contract address here to not include other transfers 
(airdrops or other stuff)

Reply:
Maybe something like this then (not sure this tuple syntax is supported in spark sql though)

```suggestion
  erc721_balances AS (
    SELECT
      holder_address,
      sum(value) AS tokens_held
    FROM
      (
        SELECT
          to AS holder_address,
          count(evt_tx_hash) AS value
        FROM
          {{ source('erc721_ethereum','evt_transfer') }}
        WHERE
          evt_block_time > '2022-04-23'
        AND (contract_address, to) IN (
            SELECT
              (nftcontractaddress, pair_address)
            FROM
              pairs_created
          )
        GROUP BY
          1
        union all
        SELECT
        FROM
          AS holder_address,
          count(evt_tx_hash) * -1 as value
        FROM
          {{ source('erc721_ethereum','evt_transfer') }}
        WHERE
          evt_block_time > '2022-04-23'
        AND (contract_address, from)
          IN (
            SELECT
              (nftcontractaddress, pair_address)
            FROM
              pairs_created
          )
        GROUP BY
          1
      ) a
    GROUP BY
      1
  ),
```